src/batch.js

```diff
@@ -26,24 +27,89 @@ export const createBatchingByteGetter = (getBytes, ranges, options) => {
-    const source = await getBytes([offset, batch[batch.length - 1][1]])
+    const source = await getBytes([batch[0][0], batch[batch.length - 1][1]])
+    consumeSource(source, batch, requests)
+    return requests[range.toString()].promise
+  }
+}
 
-    const buffer = new Uint8ArrayList()
-    await source.pipeTo(new WritableStream({ write: chunk => { buffer.append(chunk) } }))
+/**
+ * @template {unknown[]} A
+ * @template {*} T
+ * @template {*} This
+ * @param {string} spanName
+ * @param {(this: This, ...args: A) => Promise<T>} fn
+ * @param {This} [thisParam]
+ */
+export const withSimpleSpan = (spanName, fn, thisParam) =>
+  /**
+   * @param {A} args
+  */
+  async (...args) => {
+    const tracer = trace.getTracer('public-bucket')
+    const span = tracer.startSpan(spanName)
+    const ctx = trace.setSpan(context.active(), span)
 
-    for (const r of batch) {
-      requests[r.toString()].resolve(new ReadableStream({
-        pull (controller) {
-          controller.enqueue(buffer.subarray(r[0] - offset, (r[1] + 1) - offset))
-          controller.close()
-        }
-      }))
+    try {
+      const result = await context.with(ctx, fn, thisParam, ...args)
+      span.setStatus({ code: SpanStatusCode.OK })
+      span.end()
+      return result
+    } catch (err) {
+      if (err instanceof Error) {
+        span.setStatus({
+          code: SpanStatusCode.ERROR,
+          message: err.message
+        })
+      } else {
+        span.setStatus({
+          code: SpanStatusCode.ERROR
+        })
+      }
+      span.end()
+      throw err
     }
-
-    return requests[range.toString()].promise
   }
-}
+
+const consumeSource = withSimpleSpan('consumeSource',
+  /**
+ *
+ * @param {ReadableStream<Uint8Array>} source
+ * @param {import('multipart-byte-range').AbsoluteRange[]} ranges
+ * @param {Record<string, import('p-defer').DeferredPromise<ReadableStream<Uint8Array>>>} requests
+ */
+  async (source, ranges, requests) => {
+    const parts = new Uint8ArrayList()
+    // start at first byte of first blob
+    let farthestRead = ranges[0][0]
+    const offset = ranges[0][0]
+    let currentRange = 0
+    for await (const chunk of source) {
+    // append the chunk to our buffer
+      parts.append(chunk)
+      // update the absolute position of how far we've read
+      farthestRead += chunk.byteLength
+      // resolve any blobs in the current buffer
+      // note that as long as blobs are sorted ascending by start
+      // this should be resilient to overlapping ranges
+      while (farthestRead >= ranges[currentRange][1] + 1) {
+        const start = ranges[currentRange][0] - offset
+        const end = ranges[currentRange][1] + 1 - offset
+        // generate blob out of the current buffer
+        requests[ranges[currentRange].toString()].resolve(new ReadableStream({
+          pull (controller) {
+            controller.enqueue(parts.subarray(start, end))
+            controller.close()
+          }
+        }))
```

Comment:
We don't ever delete from `requests` and the resolved promises have a closure over `parts`. I think that the bytes for each batch cannot be garbage collected because of this. Can we `parts.consume()` after bytes are sent (not 100% sure if that's possible) or `delete requests[ranges[currentRange].toString()]` at the end of the stream?

I think this was an existing problem...so probably not necessary to fix here but we should resolve.

Reply:
I had some trouble when I had parts.consume in there because the pull occurs later. I think I need to parts.subarray outside the closure if we want to parts.consume.

Reply:
Could you create a new `Uint8ArrayList` from the current `parts` list in the `while` loop? You can `subarray`+`comsume` all from that when needed in the stream and `consume` from `parts` as you process the chunks.